### PR TITLE
M2 gantry fix

### PR DESCRIPTION
**Pull request by nrwslac**

### CODE REVIEW

File: HOMS_XRT/HOMS_XRT_PLC/Common/POUs/HOMS_FBs/FB_Gantry.TcPOU [hunk 1]
```diff
@@ -58,31 +58,36 @@ rtReset(CLK:=xReset);
...
+
+fbGantryDiffVirtualLimitSwitch.GantryDiff := Gantry.Diff;
+fbGantryDiffVirtualLimitSwitch.GantryDiffFlt := Gantry.DiffFlt;
+fbGantryDiffVirtualLimitSwitch(PAxis:=Gantry.PAxis, SAxis:=Gantry.SAxis);
+Gantry.PAxis.xHiLS := Gantry.PAxis.PositiveEnable;
+Gantry.PAxis.xLoLS := Gantry.PAxis.NegativeEnable;
+Gantry.SAxis.xHiLS := Gantry.SAxis.PositiveEnable;
+Gantry.SAxis.xLoLS := Gantry.SAxis.NegativeEnable;
 					
 //Direction enable logic
```
Comment (nrwslac) [inline HOMS_XRT/HOMS_XRT_PLC/Common/POUs/HOMS_FBs/FB_Gantry.TcPOU:78]: We don't care whether we're in a coupled or decouple state anymore.

File: HOMS_XRT/HOMS_XRT_PLC/Common/POUs/HOMS_FBs/FB_Gantry.TcPOU [hunk 2]
```diff
@@ -103,7 +108,7 @@ Gantry.SAxis.xEnable := (NOT Gantry.SAxis.i_xMotorInterlock) AND (Gantry.ModeReq
 //The gantry mode cannot change while anything is in motion
 	IF xFirstPass THEN
 		GC_State := GCM_Init;
-	ELSIF Gantry.PAxis.stAxis.Status.Moving OR Gantry.PAxis.stAxis.Status.Moving THEN
+	ELSIF Gantry.PAxis.stAxis.Status.Moving OR Gantry.SAxis.stAxis.Status.Moving THEN
```
Comment (nrwslac) [inline HOMS_XRT/HOMS_XRT_PLC/Common/POUs/HOMS_FBs/FB_Gantry.TcPOU:111]: Original code meant to as if secondary axis was moving?

File: HOMS_XRT/HOMS_XRT_PLC/Common/POUs/Helpers/FB_GantryDiffVirtualLimitSwitch.TcPOU [hunk 3]
```diff
@@ -32,39 +32,32 @@ X2 positive direction disabled
 X1 negative direction disabled
 *)
 IF GantryDiffFlt THEN
-	IF GantryDiff < 0  THEN
```
Comment (nrwslac) [inline HOMS_XRT/HOMS_XRT_PLC/Common/POUs/Helpers/FB_GantryDiffVirtualLimitSwitch.TcPOU:35]: This logic didn't follow the comment logic? Please double check me.

File: HOMS_XRT/HOMS_XRT_PLC/Common/POUs/Helpers/FB_GantryDiffVirtualLimitSwitch.TcPOU [hunk 4]
```diff
@@ -32,39 +32,32 @@ X2 positive direction disabled
...
-		PAxis.DecoupledNegativeEnable := FALSE;
-		SAxis.DecoupledPositiveEnable := FALSE;
-	ELSE	
-		PAxis.DecoupledNegativeEnable := TRUE;
-		SAxis.DecoupledPositiveEnable := TRUE;
-	END_IF
-	IF GantryDiff > 0 THEN
-		PAxis.DecoupledPositiveEnable := FALSE;
-		SAxis.DecoupledNegativeEnable := FALSE;
```
Comment (nrwslac) [inline HOMS_XRT/HOMS_XRT_PLC/Common/POUs/Helpers/FB_GantryDiffVirtualLimitSwitch.TcPOU:44]: Doesn't follow comment logic?

Comment (ZLLentz) [reply HOMS_XRT/HOMS_XRT_PLC/Common/POUs/Helpers/FB_GantryDiffVirtualLimitSwitch.TcPOU:44]: I'm a bit confused by this, and it's really hard to follow the diff because the variable names changed.
I'm going to sub in a number to check.

Old code with diff = 10000:
```
PAxis.NegativeEnable := TRUE;
SAxis.PositiveEnable := TRUE;
PAxis.PositiveEnable := FALSE;
SAxis.NegativeEnable := FALSE;
```
So primary has positive disabled, and secondary has negative disabled, just like the comment. The old code was correct (which makes sense, we would have had issues far sooner if this logic was wrong...)

New code with diff = 10000:
```
PAxis.NegativeEnable := FALSE;
SAxis.PositiveEnable := FALSE;
PAxis.PositiveEnable := TRUE;
SAxis.NegativeEnable := TRUE;
```
This is the opposite result.

**Review by ZLLentz — COMMENTED** [review]
I tried to take a really close look here- we need to make sure we get this right. I'm surprised to see such a large diff, I thought this would be a really short change in the big branching part where we just pick the decoupled path always.

File: HOMS_XRT/HOMS_XRT_PLC/Common/POUs/Helpers/FB_GantryDiffVirtualLimitSwitch.TcPOU (hunk 4 above)
Comment (nrwslac) [reply HOMS_XRT/HOMS_XRT_PLC/Common/POUs/Helpers/FB_GantryDiffVirtualLimitSwitch.TcPOU:44]: These are both new code; anything without Decoupled in front of it is new code

File: HOMS_XRT/HOMS_XRT_PLC/Common/POUs/HOMS_FBs/FB_Gantry.TcPOU [hunk 5]
```diff
@@ -58,31 +58,36 @@ rtReset(CLK:=xReset);
...
+//Verify the direction of motion for each axis will not increase
+//gantry difference when in a faulted state, otherwise all enables TRUE.                   
+
+fbGantryDiffVirtualLimitSwitch.GantryDiff := Gantry.Diff;
+fbGantryDiffVirtualLimitSwitch.GantryDiffFlt := Gantry.DiffFlt;
+fbGantryDiffVirtualLimitSwitch(PAxis:=Gantry.PAxis, SAxis:=Gantry.SAxis);
+Gantry.PAxis.xHiLS := Gantry.PAxis.PositiveEnable;
+Gantry.PAxis.xLoLS := Gantry.PAxis.NegativeEnable;
+Gantry.SAxis.xHiLS := Gantry.SAxis.PositiveEnable;
+Gantry.SAxis.xLoLS := Gantry.SAxis.NegativeEnable;
```
[the remark this answers is not in the document]
Comment (nrwslac) [reply HOMS_XRT/HOMS_XRT_PLC/Common/POUs/HOMS_FBs/FB_Gantry.TcPOU:76]: If the gantry difference is greater than an allowed amount we should only allow movement to correct the gantry error, otherwise move freely. Any logic beyond this seems to get messy and complicated. Do we need more than the former? 

we should make sure this doesn't ignore the limit switches.

File: HOMS_XRT/HOMS_XRT_PLC/Common/POUs/Helpers/FB_GantryDiffVirtualLimitSwitch.TcPOU (hunk 4 above)
Comment (nrwslac) [reply HOMS_XRT/HOMS_XRT_PLC/Common/POUs/Helpers/FB_GantryDiffVirtualLimitSwitch.TcPOU:44]: This change was based only on the old code conforming to the logic of the comments before it. Please take a look at the original code and double check my efforts here.

Comment (ZLLentz) [reply HOMS_XRT/HOMS_XRT_PLC/Common/POUs/Helpers/FB_GantryDiffVirtualLimitSwitch.TcPOU:44]: No, the first one was old code. I changed the variable names because otherwise my brain can't compare the two.

Comment (ZLLentz) [reply HOMS_XRT/HOMS_XRT_PLC/Common/POUs/Helpers/FB_GantryDiffVirtualLimitSwitch.TcPOU:44]: for the GH comment record:
- I misread the comment
- the comment is wrong, it has the wrong sign for the gantry diff
